src/mem/cache/prefetch/indirect_memory_v2.hh

```diff
@@ -0,0 +1,662 @@
+                    StreamDetector stream_detector;
+                    IndirectTable indirect_table;
+
+                public:
+                    PrefetchTable(const std::size_t size,
+                            const uint64_t streaming_threshold,
+                            const uint64_t max_hit_count,
+                            const uint64_t indirect_threshold):
+                        stream_detector(size, streaming_threshold),
+                        indirect_table(size, max_hit_count, indirect_threshold)
+                    {}
+
+                    constexpr StreamDetector& getStreamDetector()
+                    {
+                        return this->stream_detector;
+                    }
+
+                    constexpr IndirectTable& getIndirectTable()
+                    {
+                        return this->indirect_table;
+                    }
+            };
+
+            /*
+             * The IPD performs the computation to find the base address and
+             * shift pairs
+             */
+            class IndirectPatternDetector
+            {
+                private:
+                    struct IPDEntry
+                    {
+                        /*
+                         * The observed index values (values from array B)
+                         * which are used to compute the candidate base
+                         * addresses
+                         */
+                        std::optional<int64_t> index1;
+                        std::optional<int64_t> index2;
+
+                        /*
+                         * The number of misses for whom this enrty has base
+                         * addresses
+                         */
+                        std::size_t recorded_misses;
+
+                        /*
+                         * This tracks whether this entry generated indirect
+                         * parameters or not
+                         */
+                        bool found_params;
+
+                        /* The list of base address candidates
+                         * This structure is populate when index1 is set
+                         * When index2 is set, IMP looks up this structure to
+                         * search for matches
+                         * The outer iterator is shift values and the inner
+                         * iterator is the number of cache miss (which IMP
+                         * assumes is due to an access to indirect array, A)
+                         * observed after index1 was set, until index2 got set.
+                         * The number of tracked cache misses, i.e. the range
+                         * of the inner iterator is equal to
+                         * num_miss_after_index
+                         */
+                        std::vector<std::vector<Addr>> base_address_candidates;
+
+                        IPDEntry()
+                        {
+                            index1 = std::nullopt;
```

Comment:
nit: in general it is better to use initializer lists or in-class initialization, for both performance and versatility: https://stackoverflow.com/questions/9903248/initializing-fields-in-constructor-initializer-list-vs-constructor-body